src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -0,0 +1,302 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.DebugConfiguration;
+using Microsoft.CodeAnalysis.LanguageServer.HostWorkspace.ProjectTelemetry;
+using Microsoft.CodeAnalysis.MSBuild;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.ProjectSystem;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Microsoft.CodeAnalysis.Threading;
+using Microsoft.CodeAnalysis.Workspaces.ProjectSystem;
+using Microsoft.Extensions.Logging;
+using Microsoft.VisualStudio.Composition;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.MSBuild.BuildHostProcessManager;
+using LSP = Roslyn.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
+
+internal abstract class LanguageServerProjectLoader
```

Comment:
I don't expect this would be an abstract type, but rather a MEF component. if nothing else, that binary log numeric suffix should probably be shared across the different users.

Reply:
Alternately we could move that shared state to a separate MEF type so that bit it shared, and that's it. I guess that'll make things simpler in terms of not having to move so much state to the ProjectToLoad type.

Reply:
I wasn't exactly sure what your suggestion was here--was it possibly that `LanguageServerProjectSystem` would take a `LanguageServerProjectLoader` as a ctor parameter, store to field, and access its members in `OpenSolution` etc?

Reply:
Yes, that was my initial thinking. But it would mean that a number of things can't be passed through right away in the constructor and would then have to move to the queue items (like the project factories and such) and that does seem a bit overkill. As best I could tell the main bit of shared state was the binlog numbering, right? It'd be ideal if those were shared across the different project system, so the ordering is still clear.

Reply:
`[Export, Shared] class BinLogNamer`? It doesn't seem too bad?

Reply:
@RikkiGibson Exactly, something like that.